### PR TITLE
Add Hypothesis testing, wring out input parsing + canonicalisation

### PR DESCRIPTION
**Pull request by CyberiaResurrection**
This PR hooks up the Hypothesis testing library, which enables _property-based_ tests:

1. For all data matching some specification.
2. Perform some operations on the data.
3. Assert something about the result.

As the starline parser has historically been less than robust, I started there.
The specification was: ```Given some starline-regex matching string, it either is cleanly rejected by parsing, or results in a well-formed Star object```.  I let that rip, and smoke _poured_ out of latent weak spots and edge cases in parsing.

I then tried another test: ```Given a regex-matching string resulting in a Star object when parsed, that Star should parse cleanly to an input line and back to the same Star object, which should then parse to the same input line``` - ie, checking that functional round trips (once initially parsed) hold.  More smoke poured out of parsing.

In order to make the starline parser a little less un-robust, I compiled whitelists of star types, trade codes, and trade-code occur-in-same-star pairs from the combined TravellerMap data as at Sep 16, 2023, then applied those.

Not quite believing the effectiveness and thoroughness of the starline-parser test, I had another go on sector object generation.

_Still_ not quite believing the results, I threw Hypothesis at wringing out the canonicalisation pass of delta debugging.
Now I'm convinced.

Looking at ```test_parse_line_to_star_and_back``` , @tjoneslo , you'll see two types of ```@example``` decorators.

The larger class, such as ```'0101 000000000000000 ???????-? 000000000000000       - - 0 000   0000V'``` are inputs that Hypothesis found and reduced, that previously violated that test's specification.  I've left them there as regression tests.

The smaller class, such as ```'2405 Uurku                B986445-9 Ni Pa (Ashdak Meshukiiba) Sa          { 0 }  (A33-2) [2437] Bc   -  - 222 8  ImDv K1 V'``` are things I tripped over when running the parsing part of a Charted Space run, that also violated that test's specification.  Likewise, I've left them there as regression tests.

### CODE REVIEW

File: PyRoute/Star.py [hunk 1]
```diff
@@ -197,50 +223,72 @@ def parse_line_into_star(line, sector, pop_code, ru_calc):
 
     @staticmethod
     def _parse_line_into_star_core(star, line, sector, pop_code, ru_calc):
```
Comment (tjoneslo) [inline PyRoute/Star.py:225]: I ended up moving this entire function into an equivalent of a ` ParseStarInput` class. It keeps all the parsing bits together, and the `Star` class is more about managing the data during processing. Consider moving this, as it may simplify and isolate the different parts of the code.

File: PyRoute/Star.py [hunk 2]
```diff
@@ -749,17 +832,50 @@ def _int_to_ehex(self, value):
...
+            star_parts.append('V')
+        if 1 == len(star_parts) and star_parts[0] in types:
+            star_parts = []
+
+        if len(star_parts) < 2:
             stars = star_parts
         else:
             for (prev, current) in zip([None] + star_parts[:-1], star_parts):
-                if prev in [None, 'V', 'IV', 'Ia', 'Ib', 'II', 'III']:
+                if prev in [None, 'V', 'IV', 'Ia', 'Ib', 'II', 'III', 'VI']:
```
Comment (tjoneslo) [inline PyRoute/Star.py:858]: This is `types` defined above.
```suggestion
                if prev is None or prev in types:
```

File: PyRoute/TradeCodes.py [hunk 3]
```diff
@@ -26,6 +26,53 @@ class TradeCodes(object):
     pcolor = {'As': '#8E9397', 'De': '#d17533', 'Fl': '#e37dff', 'He': '#ff6f0c', 'Ic': '#A5F2F3',
               'Oc': '#0094ED', 'Po': '#6a986a', 'Va': '#c9c9c9', 'Wa': '#4abef4'}
     ext_codes = {'Lt', 'Ht', 'Lg', 'Hg'}
+    weird_codes = {'{Anomaly}', '{Fuel}', '{Ringworld}', '{Rosette}'}
+    allowed_residual_codes = {'Ag', 'Ba', 'Bo', 'C0', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'Cl', 'Cw',
+                              'Cy', 'D0', 'D1', 'D2', 'D3', 'D4', 'D5', 'D6', 'D7', 'D8', 'D9', 'Dw', 'Ex', 'Fr', 'Ga',
+                              'Hi', 'In', 'Lo', 'N1', 'Na', 'Ni', 'o', 'Pa', 'Ph', 'Pi', 'Po', 'Pr', 'Ri', 'Rn', 'Rv',
+                              's', 'Sp', 'St', 'Tn', 'Za'}
```
Comment (tjoneslo) [inline PyRoute/TradeCodes.py:33]: what is `s` code here?

File: PyRoute/TradeCodes.py [hunk 4]
```diff
@@ -26,6 +26,53 @@ class TradeCodes(object):
...
+    ok_pairs = {('Ni', 'Po'), ('Ni', 'Pa'), ('Ni', 'Wa'), ('Ph', 'Ri'), ('Lo', 'Va'), ('Ni', 'Va'), ('Lo', 'Po'),
```
Comment (tjoneslo) [inline PyRoute/TradeCodes.py:34]: This seems fragile to me. It's also difficult to review. 
1. Put this into alphabetical order. Keep in mind the trade codes (in the sector files) should always be in alphabetical order. So you can have (As, Va) but never (Va, As). 
2. The only codes that need to be in the list are the `pcode` (planetary codes) list above. You don't need to include the population codes (Ba, Lo, Ni, Ph, Hi) as they can go with any of the Pcodes.

File: PyRoute/TradeCodes.py [hunk 5]
```diff
@@ -26,6 +26,53 @@ class TradeCodes(object):
...
+                ('Ic', 'Na'), ('Ic', 'Ni'), ('Ic', 'Va'), ('Na', 'Ni'), ('Na', 'Va'), ('As', 'Ni'), ('As', 'Va'),
+                ('De', 'Na'), ('He', 'Na'), ('Na', 'Po'), ('Hi', 'Pr'), ('Ga', 'Hi'), ('Ag', 'Ni'), ('Na', 'Pi'),
+                ('Ni', 'Oc'), ('Ag', 'Ga'), ("Ag", "Ri"), ("Ga", "Ni"), ("Ga", "Ri"), ("Ni", "Ri"), ("Ag", "Pi"),
+                ("He", "Lo"), ("Ag", "Pr"), ("Ni", "Pr"), ("As", "Hi"), ("As", "In"), ("As", "Na"), ("Hi", "Na"),
+                ("Hi", "Va"), ("In", "Na"), ("In", "Va"), ("De", "Ph"), ("De", "Pi"), ("He", "Ph"), ("He", "Pi"),
+                ("Na", "Ph"), ("Ph", "Pi"), ("Ph", "Po"), ("Pi", "Po"), ("Ic", "Lo"), ("Fl", "He"), ("Fl", "Ni"),
+                ("As", "Lo"), ("Pa", "Ph"), ("Pa", "Ri"), ("Pa", "Pi"), ("Hi", "Wa"), ("In", "Wa"), ("Ga", "Lo"),
+                ("Ga", "Pa"), ("Ph", "Va"), ("Pi", "Va"), ("As", "Ph"), ("As", "Pi"), ("Ga", "Pr"), ("Ic", "Pi"),
+                ("Ba", "Va"), ("Hi", "Po"), ("Hi", "Ic"), ("Ic", "In"), ("Fl", "Lo"), ("De", "Ri"), ("Pi", "Wa"),
+                ("Ga", "Ph"), ("Ba", "Fl"), ("Ba", "Lo"), ("Ba", "Ni"), ("De", "Hi"), ("De", "In"), ("He", "Hi"),
```
Comment (tjoneslo) [inline PyRoute/TradeCodes.py:45]: Ba Lo and Ba Ni are not allowed codes. It used to be the population codes would overlap and you could have "Ba Lo Ni" in your trade codes.

File: PyRoute/TradeCodes.py (hunk 4 above)
Comment (tjoneslo) [reply PyRoute/TradeCodes.py:34]: Also the `Cx` and `Dx` (where x in `0-9w`) could also be removed.

File: PyRoute/TradeCodes.py [hunk 6]
```diff
@@ -80,6 +123,103 @@ def __init__(self, initial_codes):
...
+            pop = match.group(2) if match.group(2) else 'W'
+            if homeworld + '?' in initial_codes:
+                homeworld += "?"
+                pop = '0'
+            elif homeworld + str(pop) in initial_codes:
+                homeworld += str(pop)
+            sophont = "({code: <4}){pop}".format(code=match.group(1), pop=pop)
+        sophont = sophont.replace("'", "X")
+        sophont = sophont.replace("!", "X")
+        sophont = sophont.replace(" ", "X")
```
Comment (tjoneslo) [inline PyRoute/TradeCodes.py:189]: I'm trying to remember if the Traveller Wiki does this any more. The reason for this substitution was the wiki used this substitution as a recommendation from another source to create the four character sophont codes. The classic example was the K'kree. Their four character code is "K'kr" but for the wiki it became "KXkr". 

In any case, the root use of this was to allow PyRoute to accept the data from Traveller Map. Then generate output for the Traveller Wiki. This could, without having to maintain in internal list of the short code -> full name mapping, use a template and make the wiki do the lookup in the page. We should be able to remove these three lines and simply let the wiki manage the short codes.

File: PyRoute/TradeCodes.py [hunk 7]
```diff
@@ -80,6 +123,103 @@ def __init__(self, initial_codes):
...
+        @type homeworld: string
+        @type homeworlds_found: list
+        """
+        full_name = re.sub(r'\(([^)]+)\)\d?', r'\1', homeworld)
+        homeworlds_found.append(homeworld)
+        sophont = "{code: <4}{pop}".format(code=full_name[0:4], pop='W')
+        sophont = sophont.replace("'", "X")
+        sophont = sophont.replace("!", "X")
+        sophont = sophont.replace(" ", "X")
+        self.sophont_list.append(sophont)
```
Comment (tjoneslo) [inline PyRoute/TradeCodes.py:209]: Same as above.

File: PyRoute/TradeCodes.py [hunk 8]
```diff
@@ -80,6 +123,103 @@ def __init__(self, initial_codes):
...
+        self.homeworld_list.append(sophont)
+        self.codes = [code for code in self.codes if code != '[' + homeworld + "]"]
+
+    def _process_deadworld(self, deadworld, homeworlds_found):
+        full_name = re.sub(r'\(([^)]+)\)\d?', r'\1', deadworld)
+        homeworlds_found.append(deadworld)
+        sophont = "{code: <4}{pop}".format(code=full_name[0:4], pop='X')
+        sophont = sophont.replace("'", "X")
+        sophont = sophont.replace("!", "X")
+        sophont = sophont.replace(" ", "X")
```
Comment (tjoneslo) [inline PyRoute/TradeCodes.py:219]: Same as above.

File: Tests/Hypothesis/Position/testHexGubbins.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+        nu_hex = Hex.parse_from_axial(cand_hex.q, cand_hex.r)
+        self.assertIsNotNone(nu_hex)
+
+        inputs = str((pos_string, sector_x, sector_y))
+
+        self.assertEqual(cand_hex, nu_hex, str(cand_hex) + " vs " + str(nu_hex) + '\nHypothesis input:' + inputs)
+
+    def choose_sector(self, sector_choice):
+        if 0 == sector_choice:
```
Comment (tjoneslo) [inline Tests/Hypothesis/Position/testHexGubbins.py:117]: I would put this into a dict in the `__init__` function 

```
sectors = {0:  Sector(' Core', ' 0, 0'), 
 1:  Sector(' Lish', ' 0, 1'),
 2:  Sector(' Dagu', ' -1, 0')
 3: Sector(' Forn', ' 1, 0')
 4:  Sector(' Mass', ' 0, -1')}
```

This then is a simple dict lookup.

**Review by tjoneslo — COMMENTED** [review]
A first round of comments for the PR. I've not gone through all the test files, but some of the requested updates will change the tests.

File: PyRoute/DeltaStar.py [hunk 10]
```diff
@@ -11,12 +11,14 @@
...
 
     @staticmethod
-    def reduce(starline, drop_routes=False, drop_trade_codes=False, drop_noble_codes=False, drop_base_codes=False, drop_trade_zone=False, drop_extra_stars=False, reset_pbg=False, reset_worlds=False, reset_port=False, reset_tl=False, reset_sophont=False, reset_capitals=False):
+    def reduce(starline, drop_routes=False, drop_trade_codes=False, drop_noble_codes=False, drop_base_codes=False, drop_trade_zone=False, drop_extra_stars=False, reset_pbg=False, reset_worlds=False, reset_port=False, reset_tl=False, reset_sophont=False, reset_capitals=False, canonicalise=False):
         sector = Sector("dummy", " 0, 0")
         star = DeltaStar.parse_line_into_star(starline, sector, 'fixed', 'fixed')
         if not isinstance(star, DeltaStar):
             return None
 
+        if canonicalise:
```
[the remark this answers is not in the document]
Comment (CyberiaResurrection) [reply PyRoute/DeltaStar.py:20]: What would qualify to be in that report?  Starlines that aren't in canonical form?

File: PyRoute/Galaxy.py [hunk 11]
```diff
@@ -234,6 +280,65 @@ def find_world_by_pos(self, pos):
                 return world
         return None
 
+    def is_well_formed(self):
+        msg = ""
+        # check name
+        if self.name is None or '' == self.name.strip():
```
[the remark this answers is not in the document]
Comment (CyberiaResurrection) [reply PyRoute/Galaxy.py:286]: Refused to allow creation, yes.  Don't think Galaxy.name is similarly guarded.

File: PyRoute/Inputs/ParseStarInput.py [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+    @staticmethod
+    def _unpack_extension_chunks(bitz, ix_matches, ix_residual):
+        ix_data = ix_matches.groups()
+        # ix_data looks to be a little trickier, so the postprocessing is likewise
+        split_pos = len(ix_data[0])
+        bitz.append(ix_data[0])  # combined Ix, Ex, Cx - extracted to maintain compatibility with original regex
+        bitz.append(ix_data[1])  # star Ix
+        bitz.append(ix_data[2])  # star Ex
+        bitz.append(ix_data[3])  # star Cx
+        bitz.append(None)  # not sure why these are here, added to maintain compatibility with original
```
[the remark this answers is not in the document]
Comment (CyberiaResurrection) [reply PyRoute/Inputs/ParseStarInput.py:122]: You might want to add that as an issue.  IIRC, I had to add the ParseStarInput class (and all its messy internals) to ensure all 60,024 starlines in the TravMap data could be cleanly parsed or rejected.

File: PyRoute/Star.py [hunk 13]
```diff
@@ -197,50 +223,72 @@ def parse_line_into_star(line, sector, pop_code, ru_calc):
...
     @staticmethod
     def _parse_line_into_star_core(star, line, sector, pop_code, ru_calc):
+        line = line.replace('\n', '')
         star.sector = sector
         star.logger.debug(line)
-        # Cache regex lookup to avoid doing it once for check, and again to extract data
-        matches = Star.starline.match(line)
-        if matches:
-            data = matches.groups()
-        elif '{Anomaly}' in line:
```
[the remark this answers is not in the document]
Comment (CyberiaResurrection) [reply PyRoute/Star.py:206]: I didn't touch Anomaly processing, so status quo prevails there.  How would you want to process such anomalous lines?

File: PyRoute/Star.py (hunk 1 above)
Comment (CyberiaResurrection) [reply PyRoute/Star.py:225]: Good catch.

File: PyRoute/TradeCodes.py (hunk 3 above)
Comment (CyberiaResurrection) [reply PyRoute/TradeCodes.py:33]: No bloody idea.   To make that list, it turned up at least once in the TravMap data and blew up parsing.

File: PyRoute/TradeCodes.py (hunk 4 above)
Comment (CyberiaResurrection) [reply PyRoute/TradeCodes.py:34]: Again, this was summarising existing practice in the TravMap data, for much the same reason as adding ParseStarInput.

File: PyRoute/TradeCodes.py (hunk 5 above)
Comment (CyberiaResurrection) [reply PyRoute/TradeCodes.py:45]: Summarising existing practice, as above.  If a given code pair turned up at least once in the TravMap data, it got whitelisted here.

File: Tests/Hypothesis/Position/testHexGubbins.py (hunk 9 above)
Comment (CyberiaResurrection) [reply Tests/Hypothesis/Position/testHexGubbins.py:117]: That will simplify things.  Thanks.

File: PyRoute/DeltaStar.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (CyberiaResurrection) [reply PyRoute/DeltaStar.py:20]: BTW, this gubbins _is_ used, in the canonicalisation subpass of delta debugging.

File: PyRoute/Star.py (hunk 2 above)
Comment (CyberiaResurrection) [reply PyRoute/Star.py]: Fixed.  Thanks for spotting this.

File: PyRoute/Star.py (hunk 1 above)
Comment (CyberiaResurrection) [reply PyRoute/Star.py]: Your comment tripped a little reorganisation.

The old `ParseStarInput` class contained only fallback parsing, so I renamed it to `ParseStarInputFallback` (now it does what it says on the tin).

The core of the primary starline parsing, plus supporting arrays, regex, etc, is now in a new `ParseStarInput` class.

File: PyRoute/TradeCodes.py (hunk 6 above)
Comment (CyberiaResurrection) [reply PyRoute/TradeCodes.py:189]: As this was triplicated (see your later comments), I've pulled this into its own little method, namely:

```
    def _process_sophont_homeworld(self, code, pop, direct_homeworld=False):
        if direct_homeworld:
            sophont = "({code: <4}){pop}".format(code=code, pop=pop)
        else:
            sophont = "{code: <4}{pop}".format(code=code, pop=pop)
        sophont = sophont.replace("'", "X")
        sophont = sophont.replace("!", "X")
        sophont = sophont.replace(" ", "X")
        self.sophont_list.append(sophont)
        self.homeworld_list.append(sophont)

        return sophont
```

File: PyRoute/TradeCodes.py (hunk 7 above)
Comment (CyberiaResurrection) [reply PyRoute/TradeCodes.py:209]: Same as above.

File: PyRoute/TradeCodes.py (hunk 8 above)
Comment (CyberiaResurrection) [reply PyRoute/TradeCodes.py:219]: Same as above.

File: PyRoute/Galaxy.py [hunk 14]
```diff
@@ -195,14 +195,45 @@ def set_bounding_subsectors(self):
...
+            pos_bits[0] = ''
+
+        pos_trim, pos_bits[1] = self._trim_position(head_list, pos_bits[1], valid_list)
+        if pos_trim or '-' == pos_bits[1] or '+' == pos_bits[1]:
+            pos_bits[1] = ''
+
+        if 0 == len(pos_bits[0]) or 0 == len(pos_bits[1]):
+            raise ValueError("Position string malformed")
+
+        super(Sector, self).__init__(name)
```
[the remark this answers is not in the document]
Comment (CyberiaResurrection) [reply PyRoute/Galaxy.py:232]: How does position even _apply_ to Allegiance or Galaxy?

Likewise, name checks are somewhat tricky - there's enough unnamed subsectors in the TravMap data to trip up length checks on supplied name.

File: PyRoute/TradeCodes.py (hunk 4 above)
Comment (CyberiaResurrection) [reply PyRoute/TradeCodes.py]: `TradeCode.ok_pairs` is now alphabetised.

File: Tests/Hypothesis/Position/testHexGubbins.py (hunk 9 above)
Comment (CyberiaResurrection) [reply Tests/Hypothesis/Position/testHexGubbins.py]: That's fixed up - I ended up using the full sector names
